vendor/bat-native-ledger/src/bat/ledger/internal/core/shared_future.h

```diff
@@ -0,0 +1,176 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_SHARED_FUTURE_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_SHARED_FUTURE_H_
+
+#include <list>
+#include <type_traits>
+#include <utility>
+
+#include "bat/ledger/internal/core/future.h"
+
+namespace ledger {
+
+template <typename T>
+class SharedFutureState
+    : public base::RefCountedThreadSafe<SharedFutureState<T>> {
+ public:
+  SharedFutureState() : task_runner_(base::SequencedTaskRunnerHandle::Get()) {}
+
+  void SetValue(T value) {
+    task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(&SharedFutureState::SetValueInTask, this,
+                                  std::move(value)));
+  }
+
+  void AddListener(base::OnceCallback<void(const T&)> on_complete) {
+    task_runner_->PostTask(
+        FROM_HERE,
+        base::BindOnce(
+            &SharedFutureState::AddListenerInTask, this,
+            Listener{.on_complete = std::move(on_complete),
+                     .task_runner = base::SequencedTaskRunnerHandle::Get()}));
+  }
+
+ private:
+  friend class base::RefCountedThreadSafe<SharedFutureState>;
+
+  ~SharedFutureState() = default;
+
+  struct Listener {
+    base::OnceCallback<void(const T&)> on_complete;
+    scoped_refptr<base::SequencedTaskRunner> task_runner;
+  };
+
+  void AddListenerInTask(Listener listener) {
+    if (value_) {
+      listener.task_runner->PostTask(
+          FROM_HERE, base::BindOnce(RunCompleteCallback, *value_,
+                                    std::move(listener.on_complete)));
+    } else {
+      listeners_.push_back(std::move(listener));
+    }
+  }
+
+  void SetValueInTask(T value) {
+    if (value_) {
+      return;
+    }
+
+    value_ = std::move(value);
+
+    for (auto& listener : listeners_) {
+      listener.task_runner->PostTask(
+          FROM_HERE, base::BindOnce(RunCompleteCallback, *value_,
+                                    std::move(listener.on_complete)));
+    }
+
+    listeners_.clear();
+  }
+
+  static void RunCompleteCallback(
+      const T& value,
+      base::OnceCallback<void(const T&)> on_complete) {
+    std::move(on_complete).Run(value);
+  }
+
+  absl::optional<T> value_;
+  std::list<Listener> listeners_;
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+};
+
+// Like Future<T>, except that it is is copyable and the result of the
+// asynchronous operation is revealed to continuations as a const reference.
+// Always prefer |Future| to |SharedFuture|. In general, |SharedFuture| should
+// only be used when the result needs to be cached or deduped.
+//
+// Example:
+//   Future<int> future = MakeReadyFuture(42);
+//   SharedFuture shared_future(std::move(future));
+//   shared_future.Then(base::BindOnce([](const int& value) {}));
+template <typename T>
+class SharedFuture {
```

Comment:
I know I commented on this exact class in another PR or in DM specifically about not adding new callback abstractions